File: Screen.hs
module Screen where

import Text.Printf
import System.IO
import qualified Terminal
import System.IO
import System.IO.Unsafe (unsafeDupablePerformIO)

-- um pixel representa uma string que sera tratada
-- como uma unidade na tela
data IPixel = Pixel [Char]

pixelWidth = 2 -- tamanho do pixel padrao da tela

termSize = unsafeDupablePerformIO Terminal.getTermSize
termHeight = fst termSize
termWidth = snd termSize

-- tamanho da tela composta por pixels
width = termWidth `div` pixelWidth
height = termHeight

-- alguns pixels padroes
emptyPxl = replicate pixelWidth ' '

-- cria uma matrix de strings que serve como buffer para imprimir na tela
createScreenBuffer :: Int -> Int -> [Char] -> [[IPixel]]
createScreenBuffer w h c = createScreenBufferColored w h c

-- cria uma matrix de strings colorida
createScreenBufferColored :: Int -> Int -> [Char] -> [[IPixel]]
createScreenBufferColored w h c = replicate h $ replicate w $ (Pixel c)

-- largura de um buffer, maior largura entre todas as linhas
bufferWidth :: [[IPixel]] -> Int
bufferWidth buffer = maximum $ map length buffer

bufferHeight :: [[IPixel]] -> Int
bufferHeight buffer = length buffer

-- imprime um buffer no centro de outro com um deslocamento
renderCentralized :: [[IPixel]] -> [[IPixel]] -> Int -> Int -> [[IPixel]]
renderCentralized buffer source offsetX offsetY = do
    let x = (bufferWidth buffer) `div` 2 - (bufferWidth source) `div` 2 + offsetX
    let y = (bufferHeight buffer) `div` 2 - (bufferHeight source) `div` 2 + offsetY
    renderInBuffer buffer source x y

-- imprime buffer no buffer
renderInBuffer :: [[IPixel]] -> [[IPixel]] -> Int -> Int -> [[IPixel]]
renderInBuffer buffer source x y
    | length buffer == 0    = buffer
    | length source == 0    = buffer
    | y <= 0                = renderInBufferRow bufferFirst sourceFirst x : renderInBuffer bufferTail sourceTail x y
    | otherwise             = bufferFirst : renderInBuffer bufferTail source x (pred y)
    where bufferFirst = head buffer
          bufferTail = tail buffer 
          sourceFirst = head source
          sourceTail = tail source

-- imprime string em outra string
renderInBufferRow :: [IPixel] -> [IPixel] -> Int -> [IPixel]
renderInBufferRow target source i 
    | length source == 0= target
    | i >= length target= target
    | i <= 0            = sourceFirst : renderInBufferRow targetTail sourceTail 0
    | otherwise         = targetFirst : renderInBufferRow targetTail source (pred i) 
    where targetFirst = head target
          targetTail = tail target
          sourceFirst = head source
          sourceTail = tail source

-- converte string em buffer
stringToBuffer :: [Char] -> Int -> [IPixel]
stringToBuffer "" _ = []
stringToBuffer source step = do
    let rest = drop step source
    let content = take step $ source ++ (repeat ' ')
    (Pixel content) : stringToBuffer rest step

-- cria uma matrix de pixels a partir de uma lista de strings
createBufferFromStringMatrix :: [[Char]] -> [[IPixel]]
createBufferFromStringMatrix [] = []
createBufferFromStringMatrix (row:source) = do
    stringToBuffer row pixelWidth : createBufferFromStringMatrix source

bufferToString :: [[IPixel]] -> [Char]
bufferToString [] = ""
bufferToString (row:buffer) = bufferRowToString row ++ bufferToString buffer

bufferRowToString :: [IPixel] -> [Char]
bufferRowToString [] = "\n"
bufferRowToString ((Pixel content):row)= do
    content ++ bufferRowToString row

-- renderiza um buffer na tela utilizando buffer na saída do terminal
printScreen :: [[IPixel]] -> IO()
printScreen buffer = do
    hSetBuffering stdout (BlockBuffering Nothing)
    putStrLn $ bufferToString buffer
    hFlush stdout
    hSetBuffering stdout LineBuffering

    

File: Prolog/Screen.pl
% padrão de pixel
emptyPxl(' ').

% largura do pixel
pixelWidth(2).

% largura do terminal
termWidth(W):- tty_size(_,W).

% altura do terminal
termHeight(H):- tty_size(H,_).

% largura da tela em pixels
width(X) :- 
    termWidth(TW),
    pixelWidth(PW), 
    X is div(TW, PW).

% altura da tela em pixels
height(Y) :- termHeight(Y).

% cria um pixel a partir de um character
pixel(C, X) :- 
    pixelWidth(PW),
    replicate(C, PW, L),
    atomic_list_concat(L, X).

% cria um buffer a partir da largura, da altura e do conteudo do pixel
createScreenBuffer(W, H, C, Buf) :-
    pixel(C, Pxl),
    replicate(Pxl, W, Row), 
    replicate(Row, H, Buf).

% retorna o tamanho da maior linha
bufferWidth(Buf, W):- 
    maplist(length, Buf, Ls), 
    max_list(Ls, W).

% retorna o numero de linhas
bufferHeight(Buf, H):- length(Buf, H).

% retorna as dimensoes de um buffer
bufferSize(Buf, W, H):- 
    bufferWidth(Buf, W), 
    bufferHeight(Buf, H).

% renderiza um buffer dentro de outro com posicao relativa ao centro
renderCentralized(Src, Tgt, RowOff, ColOff, R):-
    bufferSize(Src, SW, SH), 
    bufferSize(Tgt, TW, TH), 
    Row is TH // 2 - SH // 2 + RowOff, 
    Col is TW // 2 - SW // 2 + ColOff, 
    renderInBuffer(Src, Tgt, Row, Col, R).
   
% renderiza um buffer dentro de outro
renderInBuffer(_, [], _, _, []):- !.
renderInBuffer([], Tgt, _, _, Tgt):- !.
renderInBuffer([SRow|Scr], [TRow|Tgt], 0, Col, R):-
    renderInBufferRow(SRow, TRow, Col, RRow), 
    renderInBuffer(Scr, Tgt, 0, Col, Rest), 
    append([RRow], Rest, R), !.
renderInBuffer(Src, [TRow|Tgt], Row, Col, R):- 
    NR is Row - 1, 
    renderInBuffer(Src, Tgt, NR, Col, Rest),
    append([TRow], Rest, R), !.

renderInBufferRow(_, [], _, []):- !.
renderInBufferRow([], Tgt, _, Tgt):- !.
renderInBufferRow([SPxl|Scr], [_|Tgt], 0, R):-
    renderInBufferRow(Scr, Tgt, 0, Rest), 
    append([SPxl], Rest, R), !.
renderInBufferRow(Scr, [TPxl|Tgt], Col, R):-
    NC is Col - 1, 
    renderInBufferRow(Scr, Tgt, NC, Rest), 
    append([TPxl], Rest, R), !.

createBufferFromStringMatrix(Matrix, Buf):-
    maplist(stringToBufferRow, Matrix, Buf).

stringToBufferRow("", []):- !.
stringToBufferRow(Text, RowList):-
    pixelWidth(PW), 
    string_length(Text, TL), 
    TL >= PW, 
    sub_string(Text, 0, PW, _, Pxl), 
    sub_string(Text, PW, _, 0, StrRest), 
    stringToBufferRow(StrRest, Rest), 
    append([Pxl], Rest, RowList), !.
stringToBufferRow(Text, RowList):-
    pixelWidth(PW), 
    string_length(Text, TL), 
    Comp is PW - TL, 
    create_string(' ', Comp, CompStr), 
    atom_concat(Text, CompStr, CompText), 
    stringToBufferRow(CompText, RowList).

% imprime buffer na tela
printScreen(Buf):-
    maplist(atomic_list_concat, Buf, RowList),
    atomic_list_concat(RowList, '\n', BufStr),
    write(BufStr).

replicate(X, N, L) :-
    length(L, N),
    maplist(=(X), L).
    
create_string(X, N, Str):- 
    replicate(X, N, L),
    atomic_list_concat(L, Str).

waitKey(ValidKeys, ReturnedKey) :-
    get_single_char(X),
    char_code(Y, X),
    (member(Y, ValidKeys) ->
    ReturnedKey = Y;
    waitKey(ValidKeys, K), ReturnedKey = K).

printArrow([],_,[]).
printArrow([Row|Rest],I,R):-
    (I =:= 0, 
     string_concat(Row," <<<",X),
     string_concat(" ", X, Z),
     R = [Z|Rest];
     N is I-1,
     I >=0,
     printArrow(Rest, N, Z),
     R = [Row|Z]).
